### ISSUE TITLE
algo: Kadane algorithm to resolve subarray max sum

### ISSUE THREAD

**Issue by blade-sensei** [blade-sensei/learning#69]

**Comment by blade-sensei** [comment]
Depends on the specific conditions be have 2 GLOBAL ways to resolve tihs. 
- If the problem ask to handle NON empty array
 - we should ask what to return if subarray is empty. 0 ? -Infinty
- If the problem doesnt ask to handle empty array, There is many ways of approch the problem. 
## 1 of  approch

Consideres that for each iteration we compute the MAX_SUM unti index [i].
```
[ 2,  3]  --> index 1 (3) 
|--|
|------| = 
|--| + |3| (i)
```
First To do that we need to SUM of the contiguous subarray[i],
We should not forget that [i] is also a subarray (of only 1 element). So we compute the the maximal the current_sum and the current element [i] (line 1 below). Now we know which if the subarray_sum is max (current_sum or [i]. But we still don't compare with [previous_sum] (subarray[i -1]) that can be greater than the current_sum. So compute this ` globalMax = Math.max(globalMax, currentMax);`
if this is first element or array of single element. the previousGlobalmax will be -infinity, to return the single element but in case it's negative it will works. Now this result is returned, globalMax will also be used for the next subarray[+1, +2..] 

### Important ❌ 
In summary ech iteration will have 3 the options of MAX_SUM at this [index]
- keep subarray[i+1] (current) ?
- keep [i] ? ```currentMax = Math.max(currentMax + number, nuber);```
- keep the subarray[i-1] ?  ```globalMax = Math.max(globalMax, currentMax);```

```math
//The expression: 
SUM_CURRENT = max_current_subarray[i-1] + [i]
max_current_subarray[i] = Max(SUM_CURRENT, [i])
```

```
//final code first approch

var maxSubArray = function(nums) {
  let currentMax = 0;
  let globalMax = -Infinity
  
  for (let number of nums) {
    currentMax = Math.max(currentMax + number, number);
    globalMax = Math.max(globalMax, currentMax);
  }
  return globalMax
};
```

**Comment by blade-sensei** [comment]
# Summary

Kadane Algorithm is used to get the maximal sum of contiguous subarray.

Basically we need to compare EACH sum of subarrays.

## What is a **a contigious subarray** respects these criterias 
-  is a part of array that compose the MAIN/Principal array
- must be composed by subarrays/elements that are next each other
- subarray can also contain a single element like [10].

. Example:
```
principal array --> [-2, 10, 2, -12]

> one of contiguous subarray is [2, -12]
 - is its part of the principal array
 - is it composed by [2] [-12] that are next each other

> [10, -12] is a non contiguous subarray: don't respect second criteria
```

## Understand the idea

## Core idea (see dynamic programming)
The main idea behind Kadane algorithm is to resolve smallest problem by one one. So the idea:
- to iterate the array from start (index 0) to the last index to resolve the main problem
- each iteration acts like it was the main problem, we store the result and other necessary state for the next index.
- the next iteration will access to the previous value to compute his own problem.

## Implement the idea with Kadane/ Which states we save

- We need to store an accumulated sum for the current level of subarray (levels will we explained further in a example). 
- And also the global_max_sum computed by compare sum levels. 

## Here for each iteration 
-  the current subarray (index) uses the previous subarray (just behind -> it's important to create contiguous) to create a new subarray. 
- this current subarray will used to check which of the 
  + recently subarray[current] 
  + and the subarray sum of previous level (or in other words the previous global_track_sum stored) is greater (MAX), the result is save in a global_track_sum.

> as a subarray can also be composed by one element like [10]
> an index = subarray[index]

```js
/**
 * if [-2, 10, 2, -12] 
 *      >
* In a classic brute force we change device the contiguous subarray this way:
* 1 -> [-2] [-2,10] [-2,10, 2] [-2,10,2,-12]
*2 -> [10]  [10,2]    [10,2,-12] 
*3 -> [2]   [2,-12]
*4 -> [-12]
* 
* We can use the same structure for Kadane algorith we considered that we have 4 POTENTIALY max sum of each LEVEL.

```

**Comment by blade-sensei** [comment]
![download_resize_jpg (1)](https://user-images.githubusercontent.com/14896541/79044841-f4afd200-7c07-11ea-95eb-25c16c301202.jpeg)
